File: typescript/src/logger.ts
/**
 * Structured logging for loop engineering
 */

import * as fs from 'fs/promises';
import * as path from 'path';

export class LoopLogger {
  private logsDir: string;

  constructor(logsDir: string) {
    this.logsDir = logsDir;
  }

  private async ensureDir() {
    await fs.mkdir(this.logsDir, { recursive: true });
  }

  private async writeJson(filename: string, entry: Record<string, unknown>) {
    await this.ensureDir();
    const logFile = path.join(this.logsDir, filename);
    await fs.appendFile(logFile, JSON.stringify(entry) + '\n', 'utf-8');
  }

  private entry(
    level: string,
    event: string,
    loopName?: string,
    extra?: Record<string, unknown>
  ): Record<string, unknown> {
    return {
      timestamp: new Date().toISOString(),
      level,
      event,
      loop_name: loopName ?? null,
      ...extra,
    };
  }

  async info(event: string, loopName?: string, extra?: Record<string, unknown>) {
    const entry = this.entry('INFO', event, loopName, extra);
    await this.writeJson('loop-engineering.log', entry);
    console.error(`[${loopName ?? 'system'}] ${event}`);
  }

  async warning(event: string, loopName?: string, extra?: Record<string, unknown>) {
    const entry = this.entry('WARNING', event, loopName, extra);
    await this.writeJson('loop-engineering.log', entry);
    console.error(`[WARN][${loopName ?? 'system'}] ${event}`);
  }

  async error(event: string, loopName?: string, extra?: Record<string, unknown>) {
    const entry = this.entry('ERROR', event, loopName, extra);
    await this.writeJson('loop-engineering.log', entry);
    console.error(`[ERROR][${loopName ?? 'system'}] ${event}`);
  }

  async runLog(loopName: string, runId: string, data: Record<string, unknown>) {
    const entry = this.entry('INFO', 'loop_run', loopName, { run_id: runId, ...data });
    await this.writeJson(`${loopName}-runs.log`, entry);
  }
}

File: python/src/loop_engineering_mcp/logger.py
"""Structured logging for loop engineering."""

import json
import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional


class LoopLogger:
    """Writes structured JSON logs to .loop/logs/."""

    def __init__(self, logs_dir: Path):
        self.logs_dir = logs_dir
        self.logs_dir.mkdir(parents=True, exist_ok=True)
        self._logger = logging.getLogger("loop_engineering")
        if not self._logger.handlers:
            handler = logging.StreamHandler()
            handler.setFormatter(logging.Formatter("%(asctime)s [%(levelname)s] %(message)s"))
            self._logger.addHandler(handler)
            self._logger.setLevel(logging.INFO)

    def _write_json(self, filename: str, entry: dict[str, Any]) -> None:
        log_file = self.logs_dir / filename
        with open(log_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry, default=str) + "\n")

    def _entry(
        self,
        level: str,
        event: str,
        loop_name: Optional[str] = None,
        **extra: Any,
    ) -> dict[str, Any]:
        return {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": level,
            "event": event,
            "loop_name": loop_name,
            **extra,
        }

    def info(self, event: str, loop_name: Optional[str] = None, **extra: Any) -> None:
        entry = self._entry("INFO", event, loop_name, **extra)
        self._write_json("loop-engineering.log", entry)
        self._logger.info(f"[{loop_name or 'system'}] {event}")

    def warning(self, event: str, loop_name: Optional[str] = None, **extra: Any) -> None:
        entry = self._entry("WARNING", event, loop_name, **extra)
        self._write_json("loop-engineering.log", entry)
        self._logger.warning(f"[{loop_name or 'system'}] {event}")

    def error(self, event: str, loop_name: Optional[str] = None, **extra: Any) -> None:
        entry = self._entry("ERROR", event, loop_name, **extra)
        self._write_json("loop-engineering.log", entry)
        self._logger.error(f"[{loop_name or 'system'}] {event}")

    def run_log(self, loop_name: str, run_id: str, data: dict[str, Any]) -> None:
        entry = self._entry("INFO", "loop_run", loop_name, run_id=run_id, **data)
        self._write_json(f"{loop_name}-runs.log", entry)
